File: cpp/leetcode/099.cpp
class Solution {
public:
    void recoverTree(TreeNode *root) {
        TreeNode *first = NULL, *second = NULL, *ptr = NULL;
        help(root, ptr, first, second);
        swap(first->val, second->val); //swap value
    }

    void help(TreeNode *root, TreeNode *&ptr, TreeNode *&first, TreeNode *&second) {
        if (!root) {
            return;
        }
        help(root->left, ptr, first, second);
        if (!first && ptr && ptr->val >= root->val) {
            first = ptr;
        }
        if (first && ptr && ptr->val >= root->val) {
            second = root;
        }
        ptr = root;
        help(root->right, ptr, first, second);
    }
};

File: python/leetcode/099.py
class Solution(object):
    def recoverTree(self, root):
        if not root:
            return
        self.first, self.second = None, None
        self.pre = TreeNode(-2**31)
        self.help(root)
        self.first.val, self.second.val = self.second.val, self.first.val

    def help(self, root):
        if not root:
            return
        self.help(root.left)
        if not self.first and self.pre.val >= root.val:
            self.first = self.pre
        if self.first and self.pre.val >= root.val:
            self.second = root
        self.pre = root
        self.help(root.right)
